### ISSUE TITLE
Android app crashes with error split_config.arm64_v8a.apk!libmonosgen-2.0.so

### ISSUE THREAD

**Issue by sunilsolanki-alight** [dotnet/runtime#109921]
### Description

split_config.arm64_v8a.apk!libmonosgen-2.0.so

### Version with bug

8.0.91 SR9.1

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.91 SR9.1

### Affected platforms

Android

### Did you find any workaround?

No

### Relevant log output

```shell
split_config.arm64_v8a.apk!libmonosgen-2.0.so
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 31471 >>> xxx.xxx.com <<<

backtrace:
  #00  pc 0x00000000002a42a4  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so 
  dotnet/maui#1  pc 0x00000000002a35bc  /data/app/~~302XoCoC298UZF5Zrmfk_w==/cxxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so
  dotnet/maui#2  pc 0x000000000029dfe4  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (
  dotnet/maui#3  pc 0x00000000002c7034  /data/app/~~302XoCoC298UZF5Zrmfk_w==/cxxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so 
  dotnet/maui#4  pc 0x00000000002c61d0  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so 
  dotnet/maui#5  pc 0x00000000002c2564  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so 
  dotnet/maui#6  pc 0x00000000002c2330  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so 
  dotnet/maui#7  pc 0x00000000002beafc  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (
  dotnet/maui#8  pc 0x00000000002befe0  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (
  dotnet/maui#9  pc 0x00000000002a6ab0  /data/app/~~302XoCoC298UZF5Zrmfk_w==/xxx.xxx.com-KrkS0UsNZ9pNu38dNjI5YQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so
```

**Comment by kg** [comment]
> Not to pile on, but I also see this issue in the Google Play Console for a production .NET MAUI Android application. Here are some additional details I see there from Google Play:
> 
> [split_config.arm64_v8a.apk!libmonosgen-2.0.so] SIGSEGV
> 
> Problem This issue includes GWP-ASan stack traces. These stack traces provide additional information for debugging. Recommendation Some stack traces below include additional details about memory allocation. They may help debug this issue.
> 
> ```
> *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
> pid: 0, tid: 21357 >>> com.******.app <<<
> 
> backtrace:
>   #00  pc 0x00000000000d75cc  /data/app/~~ro-aanPmvChnEGfYdE_A-w==/com.******.app-TtQeBmv5rm4QixhQ0dzHVw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 943992f94e0687d44fe2a2fc0648c7cc473eee93)
>   dotnet/android#1  pc 0x00000000000ca26c  /data/app/~~ro-aanPmvChnEGfYdE_A-w==/com.******.app-TtQeBmv5rm4QixhQ0dzHVw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 943992f94e0687d44fe2a2fc0648c7cc473eee93)
> ~snip~

Hi,
Can you let me know which specific version of .NET you used to build this so I can match it with the source code for that version? Thank you!

**Comment by grendello** [comment]
@kg `libmonosgen-2.0.so` attached above says `8.0.824.36612 @Commit: 08338fcaa5c9b9a8190abb99222fed12aaba956c`

**Comment by kg** [comment]
The stack trace in this case suggests heap corruption, specifically of field and/or method metadata. The asan stacktrace (thank you very much for providing it!) also suggests this. Are your application(s) making much use of pinvoke, JNI, etc? Calling into any third-party native libraries?

Thanks.

**Comment by VNGames** [comment]
A few more stack traces

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 28200 >>> com.example.app <<<

backtrace:
  #00  pc 0x00000000000d77a4  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #01  pc 0x00000000000ca444  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #02  pc 0x00000000000e3f0c  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #03  pc 0x00000000000bb2f8  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #04  pc 0x00000000000bd934  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #05  pc 0x00000000000c264c  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #06  pc 0x00000000000c1a6c  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #07  pc 0x0000000000152278  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #08  pc 0x0000000000151ddc  /data/app/~~9Zfo4zOl7WqWCiDvRydiaQ==/com.example.app-enZujLkwQVZO0lw0cRXUvA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #09  pc 0x0000000000004300 
  
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 4939 >>> com.example.app <<<

backtrace:
  #00  pc 0x00000000000d77a4  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #01  pc 0x00000000000ca444  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #02  pc 0x00000000000e3f0c  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #03  pc 0x00000000000ca444  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #04  pc 0x00000000000e3f0c  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #05  pc 0x00000000000bb2f8  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #06  pc 0x00000000000bd934  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #07  pc 0x00000000000c264c  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #08  pc 0x00000000000c1a6c  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #09  pc 0x0000000000152278  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #10  pc 0x0000000000151ddc  /data/app/~~5bKw2v4EVXMKZasEwHuqMw==/com.example.app-hTUJusGPdrz31QRSPLfm9A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #11  pc 0x0000000000004300 
  
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 20170 >>> com.example.app <<<

backtrace:
  #00  pc 0x00000000002f3798  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #01  pc 0x00000000002f360c  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #02  pc 0x00000000002f2684  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #03  pc 0x00000000002f3590  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #04  pc 0x00000000002c71e0  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #05  pc 0x00000000002c6308  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #06  pc 0x00000000002c269c  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #07  pc 0x00000000002c2468  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #08  pc 0x00000000002bec78  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #09  pc 0x00000000002a75f4  /data/app/~~ZwFMKv9ew4yqOZY-IvxbRg==/com.example.app-tQ0qzgWqFpe1DgEE5Uw14Q==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #10  pc 0x000000000000dc08
```

[libmonosgen-2.0.so.zip](https://github.com/user-attachments/files/18386151/libmonosgen-2.0.so.zip)

**Comment by alight-rajeshrathod** [comment]
> > [@grendello](https://github.com/grendello) please see the uploaded zip file containing this library. [libmonosgen-2.0.so.zip](https://github.com/user-attachments/files/17552009/libmonosgen-2.0.so.zip)
> 
> Thanks, symbolication of the trace you provided in [#109921](https://github.com/dotnet/runtime/issues/109921) is as follows:
> 
> ```
> 0xd75cc
> mono_method_to_ir
> /__w/1/s/src/mono/mono/mini/method-to-ir.c:9914
> 
> 0xca26c
> inline_method
> /__w/1/s/src/mono/mono/mini/method-to-ir.c:0
> 
> 0xebe4c
> handle_ctor_call
> /__w/1/s/src/mono/mono/mini/method-to-ir.c:5807
> mono_method_to_ir
> /__w/1/s/src/mono/mono/mini/method-to-ir.c:9364
> 
> 0xbb168
> mono_compile_create_vars
> /__w/1/s/src/mono/mono/mini/mini.c:1927
> mini_method_compile
> /__w/1/s/src/mono/mono/mini/mini.c:3492
> 
> 0xbd7a4
> mini_class_has_reference_variant_generic_argument
> /__w/1/s/src/mono/mono/mini/mini.c:4040
> 
> 0xc2470
> compile_special
> /__w/1/s/src/mono/mono/mini/mini-runtime.c:0
> mono_jit_compile_method_with_opt
> /__w/1/s/src/mono/mono/mini/mini-runtime.c:2778
> jit_compile_method_with_opt_cb
> /__w/1/s/src/mono/mono/mini/mini-runtime.c:2868
> jit_compile_method_with_opt
> /__w/1/s/src/mono/mono/mini/mini-runtime.c:2884
> 
> 0xc18dc
> fwrite(void const*, unsigned long pass_object_size0, unsigned long, __sFILE*)
> /usr/local/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/bits/fortify/stdio.h:125
> mono_emit_jit_dump
> /__w/1/s/src/mono/mono/mini/mini-runtime.c:2219
> 
> 0x152048
> common_call_trampoline
> /__w/1/s/src/mono/mono/mini/mini-trampolines.c:0
> 
> 0x151bac
> mini_add_method_trampoline
> /__w/1/s/src/mono/mono/mini/mini-trampolines.c:423
> ```
> 
> This looks like a JIT issue with some generic class, if other people provide their copy of `libmonosgen-2.0.so`, we would be able to see if the traces generally match.
> 
> [@steveisok](https://github.com/steveisok) can you take a look and route appropriately?

Can you please guide on how did you get this symbolic trace from .so file?

**Comment by steveisok** [comment]
@sunilsolanki-alight @VNGames can you please try the latest .NET 8 or 9 and see if this is still an issue? My team reminded me I may have fixed this back in November and the fix started showing up in 9.0.1 and 8.0.12 respectively.

Fix for reference: https://github.com/dotnet/runtime/pull/109381

**Comment by mnxamdev** [comment]
Note we reported it on 8.0.82 so not fixed in 8.0.12

https://github.com/dotnet/runtime/issues/109921#issuecomment-2483058413


—

Giancarlo Espinosa
Lead Mobile Software Engineer
***@***.***<https://editor-us.codetwo.com/%7BE-mail%7D>
Musicnotes.com  |  901 Deming Way #100, Madison, WI 53717

[cid:mn-logo_cc2bec9e-bfb5-458d-8a30-a9a0cd14f436.png]


________________________________

**Comment by steveisok** [comment]
> Note we reported it on 8.0.82 so not fixed in 8.0.12

Our versions are confusing, sorry. 8.0.12 is the runtime version and that goes along with 8.0.405 of the SDK.

**Comment by mnxamdev** [comment]
Thx for the clarification @steveisok we will definitely check! Thx to everyone for all their efforts!

**Comment by lateralusX** [comment]
Probably related to this https://github.com/dotnet/runtime/issues/109111 and fixed in this backport, https://github.com/dotnet/runtime/pull/109511

**Comment by lateralusX** [comment]
The ASan report https://github.com/dotnet/runtime/issues/109921#issuecomment-2483058374 looks similar to the one in https://github.com/dotnet/runtime/issues/109111. I will need to symbolicate the callstack here to be 100% sure, but looking at the issue with `try_prepare_objaddr_callvirt_optimization` aligns with this report as well. Since that code before the fix of https://github.com/dotnet/runtime/pull/109511 would assume that MonoType * would hold a MonoClass * and then  `try_prepare_objaddr_callvirt_optimization` would access the field MonoClass.klass_kind that is at offset 27, but since it's actually a MonoType that is 16 byte it will read 11 bytes to the right of the allocated block, as reported by the ASan report. In case ASan was not monitoring that allocation, this could end up in some bad optimization in `try_prepare_objaddr_callvirt_optimization` that would lead to random behaviors probably leading to other crashes.

**Comment by carzh** [comment]
I encountered a similar issue -- signed release APK crashing because of missing libmonosgen-2.0.so when loaded onto Android device. 

Mentioning what ended up resolving it for me in case it's helpful for others because I have found the process of developing in MAUI/dotnet a little frustrating so far --

Previously I had:
```
<PropertyGroup Condition="'$(IsAndroidTarget)' == 'true'">
	  <AndroidPackageFormat>apk</AndroidPackageFormat>
	  <EmbedAssembliesIntoApk>True</EmbedAssembliesIntoApk>
	  <AndroidUseAapt2>True</AndroidUseAapt2>
	</PropertyGroup>
```

Edited it to:

```
<PropertyGroup Condition="'$(IsAndroidTarget)' == 'true'">
	  <AndroidPackageFormat>apk</AndroidPackageFormat>
	</PropertyGroup>
```

I believe the issue is related to how the APK is bundled, so toggling some settings even if it's just putting the default behavior or removing tags for the default behavior may help.

**Comment by Manish-Pradhan-FP** [comment]
Seeing the same issue in release app

`*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 22136 >>> net.Fieldpoint.MobileTech <<<

backtrace:
  #00  pc 0x00000000002a42d4  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #01  pc 0x00000000002a3168  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #02  pc 0x000000000029dfe4  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #03  pc 0x00000000002c7034  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #04  pc 0x00000000002c61d0  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #05  pc 0x00000000002c2564  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #06  pc 0x00000000002c2330  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #07  pc 0x00000000002beafc  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #08  pc 0x00000000002a74bc  /data/app/~~sMZNr8xmCR7d1NjvSyCHTA==/net.Fieldpoint.MobileTech-eCUK4blRpKIEgMd7f8crnQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: eb24a52f2dd44e052b62836d7c0d12115d335fb9)
  #09  pc 0x000000000000e468 `

**Comment by alight-rajeshrathod** [comment]
We continue to have this issue in prod app despite having following fixes in the release app.

1. Upgrade to MAUI 8.0.100
2. Upgrade .net SDK runtime to 8.0.403
3. Add concurrent gc disabled flag in .csproj.

@steveisok - Any direction for this issue is highly appreciated as google playstore crash reports are creating panic in the team.

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 29497 >>> com.abc.ddddd <<<

backtrace:
  #00  pc 0x000000000021370c  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
  #01  pc 0x00000000002230b0  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
  #02  pc 0x000000000022b204  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
  #03  pc 0x000000000004819c
```

**Comment by lateralusX** [comment]
> We continue to have this issue in prod app despite having following fixes in the release app.
> 
> 1. Upgrade to MAUI 8.0.100
> 2. Upgrade .net SDK runtime to 8.0.403
> 3. Add concurrent gc disabled flag in .csproj.
> 
> [@steveisok](https://github.com/steveisok) - Any direction for this issue is highly appreciated as google playstore crash reports are creating panic in the team.
> 
> ```
> *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
> pid: 0, tid: 29497 >>> com.abc.ddddd <<<
> 
> backtrace:
>   #00  pc 0x000000000021370c  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
>   #01  pc 0x00000000002230b0  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
>   #02  pc 0x000000000022b204  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
>   #03  pc 0x000000000004819c
> ```

This looks like a different issue than whats been discussed here, when I symbolicate above we get the following stacktrace:

mono_class_setup_vtable_full at /__w/1/s/src/mono/mono/metadata/class-setup-vtable.c:900
init_io_stream_slots at /__w/1/s/src/mono/mono/metadata/icall.c:3378
ves_icall_System_IO_Stream_HasOverriddenBeginEndRead at /__w/1/s/src/mono/mono/metadata/icall.c:3437
 (inlined by) ves_icall_System_IO_Stream_HasOverriddenBeginEndRead_raw at /__w/1/s/src/mono/mono/metadata/../metadata/icall-def.h:228

So its crashing when setting up the vtable for System.IO.Stream the first time, most likely since klass is NULL. I need to look a little closer at that code, but if that happens is probably a race condition between threads on init_io_stream_slots. Do you have more threads in this crash dump that have functions in mono runtime functions?

**Comment by lateralusX** [comment]
> I encountered a similar issue -- signed release APK crashing because of missing libmonosgen-2.0.so when loaded onto Android device.
> 
> Mentioning what ended up resolving it for me in case it's helpful for others because I have found the process of developing in MAUI/dotnet a little frustrating so far --
> 
> Previously I had:
> 
> ```
> <PropertyGroup Condition="'$(IsAndroidTarget)' == 'true'">
> 	  <AndroidPackageFormat>apk</AndroidPackageFormat>
> 	  <EmbedAssembliesIntoApk>True</EmbedAssembliesIntoApk>
> 	  <AndroidUseAapt2>True</AndroidUseAapt2>
> 	</PropertyGroup>
> ```
> 
> Edited it to:
> 
> ```
> <PropertyGroup Condition="'$(IsAndroidTarget)' == 'true'">
> 	  <AndroidPackageFormat>apk</AndroidPackageFormat>
> 	</PropertyGroup>
> ```
> 
> I believe the issue is related to how the APK is bundled, so toggling some settings even if it's just putting the default behavior or removing tags for the default behavior may help.

@carzh did you get any callstacks to share for this scenario? Would be interesting to see what caused the issue so we could route it to the right team.

**Comment by lateralusX** [comment]
> 22b0fc

This is the same crash as https://github.com/dotnet/runtime/issues/109921#issuecomment-2640551190 so will have the same fix.

**Comment by lateralusX** [comment]
> Same issue here, reported in the PlayStore:
> 
> ```
> *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
> pid: 0, tid: 30178 >>> de.abc.members_app <<<
> 
> backtrace:
>   #00  pc 0x00000000000d77a4  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#1  pc 0x00000000000ca444  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#2  pc 0x00000000000ec024  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#3  pc 0x00000000000bb2f8  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#4  pc 0x00000000000bd934  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#5  pc 0x00000000000c264c  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#6  pc 0x00000000000c1a6c  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#7  pc 0x0000000000152278  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#8  pc 0x0000000000151ddc  /data/app/~~y6f1LqRn_43bgGP2uHAJpA==/de.abc.members_app-NO1P3uSWvR8hhAR80SX-sg==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
>   dotnet/android#9  pc 0x0000000000004300 
> ```
> 
> Device: Galaxy S22 Ultra with Android 14 (SDK 34)

This crash seems to show up in try_prepare_objaddr_callvirt_optimization and fix to that should be included in 8.0.12.

**Comment by alight-rajeshrathod** [comment]
> > We continue to have this issue in prod app despite having following fixes in the release app.
> > 
> > 1. Upgrade to MAUI 8.0.100
> > 2. Upgrade .net SDK runtime to 8.0.403
> > 3. Add concurrent gc disabled flag in .csproj.
> > 
> > [@steveisok](https://github.com/steveisok) - Any direction for this issue is highly appreciated as google playstore crash reports are creating panic in the team.
> > ```
> > *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
> > pid: 0, tid: 29497 >>> com.abc.ddddd <<<
> > 
> > backtrace:
> >   #00  pc 0x000000000021370c  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
> >   #01  pc 0x00000000002230b0  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
> >   #02  pc 0x000000000022b204  /data/app/~~037ewMW_xtNhrbu_9epDuA==/com.abc.ddddd-EhdN-67TFLDS8VXgaVAFUQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: c7db5d00fcd3ecbca8fb4a5d7087a9c5b3c4f43f)
> >   #03  pc 0x000000000004819c
> > ```
> 
> This looks like a different issue than whats been discussed here, when I symbolicate above we get the following stacktrace:
> 
> ```
> mono_class_setup_vtable_full at /__w/1/s/src/mono/mono/metadata/class-setup-vtable.c:900
> init_io_stream_slots at /__w/1/s/src/mono/mono/metadata/icall.c:3378
> ves_icall_System_IO_Stream_HasOverriddenBeginEndRead at /__w/1/s/src/mono/mono/metadata/icall.c:3437
>  (inlined by) ves_icall_System_IO_Stream_HasOverriddenBeginEndRead_raw at /__w/1/s/src/mono/mono/metadata/../metadata/icall-def.h:228
> ```
> 
> So its crashing when setting up the vtable for System.IO.Stream the first time, most likely since klass is NULL. I need to look a little closer at that code, but if that happens is probably a race condition between threads on init_io_stream_slots. Do you have more threads in this crash dump that have callstacks including mono runtime functions?

Thanks @lateralusX. I have found few other stacktrace as below and not sure if that helps.

`*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 24051 >>> com.abc.ddddd <<<

backtrace:
  #00  pc 0x00000000002dc160  /data/app/~~tCyRftSC-eZwC7-Pa2ZCFA==/com.abc.ddddd-97z9LyNgbQpEokW4ESKWuA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #01  pc 0x00000000002dded8  /data/app/~~tCyRftSC-eZwC7-Pa2ZCFA==/com.abc.ddddd-97z9LyNgbQpEokW4ESKWuA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #02  pc 0x00000000002fc404  /data/app/~~tCyRftSC-eZwC7-Pa2ZCFA==/com.abc.ddddd-97z9LyNgbQpEokW4ESKWuA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #03  pc 0x00000000002fb490  /data/app/~~tCyRftSC-eZwC7-Pa2ZCFA==/com.abc.ddddd-97z9LyNgbQpEokW4ESKWuA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #04  pc 0x0000000000070334  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+180)
  #05  pc 0x0000000000062060  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)`

`*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 22978 >>> com.abc.ddddd <<<

backtrace:
  #00  pc 0x00000000000d77a4  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #01  pc 0x00000000000ca444  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #02  pc 0x00000000000ec024  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #03  pc 0x00000000000ca444  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #04  pc 0x00000000000e3f0c  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #05  pc 0x00000000000bb2f8  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #06  pc 0x00000000000bd934  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #07  pc 0x00000000000c264c  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #08  pc 0x00000000000c1a6c  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #09  pc 0x0000000000152278  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #10  pc 0x0000000000151ddc  /data/app/~~_uJdJXElEP7RkUgSQNUmvA==/com.abc.ddddd-k0I-j6RtFBNa5x7g3arz4A==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 67228244379e0d7698287264cbddfcea8c1e1cb3)
  #11  pc 0x0000000000004300 `

**Comment by lateralusX** [comment]
This fix https://github.com/dotnet/runtime/pull/112282, as well as backports to net9 and net8 should take care of the crash related to this callstack:

mono_class_setup_vtable_full at /__w/1/s/src/mono/mono/metadata/class-setup-vtable.c:900
init_io_stream_slots at /__w/1/s/src/mono/mono/metadata/icall.c:3378
ves_icall_System_IO_Stream_HasOverriddenBeginEndRead at /__w/1/s/src/mono/mono/metadata/icall.c:3437
 (inlined by) ves_icall_System_IO_Stream_HasOverriddenBeginEndRead_raw at /__w/1/s/src/mono/mono/metadata/../metadata/icall-def.h:228

It should be possible to workaround this race calling to ReadAsync on Stream is done during before other threads might end up calling the same initialization in parallel.

**Comment by dbSoft-SoftwareSolutions** [comment]
We have the exactly same issue using mono android native. Net 9

**Comment by l0gaw** [comment]
I've been facing this in the .NET 9 and android in Samsung Devices. When I run a foreground service along with collection view scrolling, the app eventually crashes, and I receive this error. It doesn't happen with iOS.

![Image](https://github.com/user-attachments/assets/faf4e9ab-0720-45bb-b527-eee4d6617cd8)

**Comment by lateralusX** [comment]
@l0gaw, this is a different issue than what's been discussed here. All crashes that are not correctly symbolicated happening inside Mono VM will end up somewhere inside libmonosgen-2.0.so. When I symbolicate the above callstack I get:

0x299548: bridge_object_forward at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:391
 (inlined by) push_object at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:629
 0x2988ac: push_all at /__w/1/s/src/mono/mono/metadata/../sgen/sgen-scan-object.h:59
 (inlined by) dfs at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:886
 (inlined by) processing_stw_step at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:1019
 0x293fa0: sgen_bridge_processing_stw_step at /__w/1/s/src/mono/mono/metadata/sgen-bridge.c:224
 0x2b1e18: sgen_client_bridge_processing_stw_step at /__w/1/s/src/mono/mono/sgen/../metadata/sgen-client-mono.h:239
 (inlined by) finish_gray_stack at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:1191
 0x2b0f50: collect_nursery at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:1930
 0x2ad494: sgen_perform_collection_inner at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:2650
 (inlined by) sgen_perform_collection at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:2762
 0x2ad27c: sgen_ensure_free_space at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:2616
 0x2a9ad8: sgen_alloc_obj_nolock at /__w/1/s/src/mono/mono/sgen/sgen-alloc.c:279

so its a crash inside tarjan bridge and it appears that it has a managed object that it scans that includes a reference to 0x4000 that is an invalid object address, so the managed object internals have probably been corrupted at an early stage.

Do you have a repro that you can run locally to reproduce this issue? If so, it could be interesting to see if we could extract more information before the crash happens. It could also be interesting to eliminate the tarjan bridge out of the equation and see if it still repros, you can look at this comment on how to switch GC bridge implementation on Android, https://github.com/dotnet/runtime/issues/106410#issuecomment-2391391866.

**Comment by fabEps** [comment]
Still with NET 9.0.40, I got several crashes which are tracked on PlayStore as unstable app:

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 16530 >>> de.vpt.members_app <<<

backtrace:
  #00  pc 0x000000000046c000  /apex/com.android.art/lib64/libart.so (art::JNI<false>::GetObjectClass(_JNIEnv*, _jobject*)+600)
  #01  pc 0x00000000000209e0  /data/app/~~gWmmlG98qMiBtWKvIMVRWA==/de.vpt.members_app-nAH2wKPS3hAx6ANCjmliBg==/split_config.arm64_v8a.apk
```

- Samsung S10, Android 12 (SDK 31)
- Samsung S25, Android 15 (SDK 35)

How can I provide more information?

**Comment by mnxamdev** [comment]
Like @fabEps this is affecting our app rating and Play Store search results.

**Comment by ahoffmandevelopment** [comment]
Just checking on the progress on this. We have several android apps in the App Store that are crashing with this stack trace:

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 29124 >>> com.xxxx.xxxx <<<

backtrace:
  #00  pc 0x0000000000020990  /data/app/~~jBJ9ykZBLfRoq9luRXgNpg==/com.xxxx.xxxx-LmuHifoVLC3fAGg9SduwIQ==/split_config.arm64_v8a.apk
```

Is there a specific maui/runtime version I should be targeting? Or some change that needs made to the config?

For reference, this is what my config looks like:

```
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-android|AnyCPU'"> 
	  <AndroidDexTool>d8</AndroidDexTool> 
	  <AndroidLinkTool>r8</AndroidLinkTool>
	  <AndroidEnableSGenConcurrent>false</AndroidEnableSGenConcurrent>
	  <RunAOTCompilation>true</RunAOTCompilation> 
</PropertyGroup> 
```

I have also tried RunAOTCompilation false with no difference in crashes.

**Comment by SprengerS** [comment]
Well we have the same problem since using .net 9. This is the stack trace from play console

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 12624 >>> eu...... <<<

backtrace:
  #00  pc 0x000000000021fd04  /data/app/~~ANugfELUkafJFq7JOwEz9Q==/eu.....-yicNSgLveGxQzuyblxTSbA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 5c0f30dec901885b27282d3532d8efe2d5d00206)
  #01  pc 0x000000000022f3b8  /data/app/~~ANugfELUkafJFq7JOwEz9Q==/eu.....-yicNSgLveGxQzuyblxTSbA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 5c0f30dec901885b27282d3532d8efe2d5d00206)
  #02  pc 0x0000000000237278  /data/app/~~ANugfELUkafJFq7JOwEz9Q==/eu......-yicNSgLveGxQzuyblxTSbA==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 5c0f30dec901885b27282d3532d8efe2d5d00206)
  #03  pc 0x000000000003c4cc 
```

**Comment by fabEps** [comment]
Will this be fixed by https://github.com/dotnet/runtime/issues/106410#issuecomment-2854575035 ? Sorry, but this is getting really critical now:

![Image](https://github.com/user-attachments/assets/678a85d9-95bc-4f48-ae13-77606e6366d5)

![Image](https://github.com/user-attachments/assets/44bd9b09-2a4c-40e3-94f8-6a72f04d5985)

**Comment by ivmirx** [comment]
I also have some of those in an app using native UIs and built with .NET 9.0.3, Google Play managed to symbolicate them:

```
  #00  pc 0x0000000000542290  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::MarkingPhase()+680)
  #01  pc 0x0000000000565a60  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::RunPhases()+168)
  #02  pc 0x0000000000591b40  /apex/com.android.art/lib64/libart.so (art::gc::collector::GarbageCollector::Run(art::gc::GcCause, bool)+324)
  #03  pc 0x0000000000438220  /apex/com.android.art/lib64/libart.so (art::gc::Heap::CollectGarbageInternal(art::gc::collector::GcType, art::gc::GcCause, bool, unsigned int)+564)
  #04  pc 0x00000000003843c0  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+112)
  #05  pc 0x0000000000438880  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (java.lang.Runtime.gc+160)
  #06  pc 0x0000000000368774  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #07  pc 0x0000000000367168  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithVarArgs<art::ArtMethod*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, art::ArtMethod*, std::__va_list)+812)
  #08  pc 0x000000000072c0cc  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list)+192)
  #09  pc 0x0000000000044a04  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmono-android.release.so (_JNIEnv::CallVoidMethod(_jobject*, _jmethodID*, ...)+120) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #10  pc 0x0000000000041ab4  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmono-android.release.so (xamarin::android::internal::OSBridge::gc_cross_references(int, MonoGCBridgeSCC**, int, MonoGCBridgeXRef*)+228) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #11  pc 0x000000000029499c  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #12  pc 0x00000000002b0ee4  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #13  pc 0x00000000002adf80  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #14  pc 0x00000000002aef1c  /data/app/~~m5SBilDDpGug3AoNhoGKYg==/MyApp-5sRolc2NO023bgyFoy0eSw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
```

```
  #00  pc 0x0000000000257180  /apex/com.android.art/lib64/libart.so (art::gc::ReferenceQueue::ClearWhiteReferences(art::gc::ReferenceQueue*, art::gc::collector::GarbageCollector*, bool)+160)
  #01  pc 0x0000000000256e48  /apex/com.android.art/lib64/libart.so (art::gc::ReferenceProcessor::ProcessReferences(art::Thread*, art::TimingLogger*)+540)
  #02  pc 0x0000000000257fa0  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::ReclaimPhase()+120)
  #03  pc 0x00000000004089bc  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::RunPhases()+392)
  #04  pc 0x00000000003d3e20  /apex/com.android.art/lib64/libart.so (art::gc::collector::GarbageCollector::Run(art::gc::GcCause, bool)+336)
  #05  pc 0x000000000048a434  /apex/com.android.art/lib64/libart.so (art::gc::Heap::CollectGarbageInternal(art::gc::collector::GcType, art::gc::GcCause, bool, unsigned int)+508)
  #06  pc 0x0000000000da4790  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+112)
  #07  pc 0x0000000000389d00  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (java.lang.Runtime.gc+160)
  #08  pc 0x0000000000328194  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #09  pc 0x00000000003262f8  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithVarArgs<_jmethodID*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, std::__va_list)+924)
  #10  pc 0x0000000000632850  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list)+180)
  #11  pc 0x0000000000044a04  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmono-android.release.so (_JNIEnv::CallVoidMethod(_jobject*, _jmethodID*, ...)+120) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #12  pc 0x0000000000041ab4  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmono-android.release.so (xamarin::android::internal::OSBridge::gc_cross_references(int, MonoGCBridgeSCC**, int, MonoGCBridgeXRef*)+228) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #13  pc 0x000000000029499c  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #14  pc 0x00000000002b0ee4  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #15  pc 0x00000000002adf80  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #16  pc 0x00000000002aef1c  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #17  pc 0x000000000029e108  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (mono_gc_collect+44) (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
```

```
  #00  pc 0x00000000002491b0  /apex/com.android.art/lib64/libart.so (void art::gc::collector::MarkCompact::CompactPage<false>(art::mirror::Object*, unsigned int, unsigned char*, unsigned char*, bool)+7084)
  #01  pc 0x000000000024d9c0  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::CompactionPhase()+432)
  #02  pc 0x0000000000408ab4  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::RunPhases()+640)
  #03  pc 0x00000000003d3e20  /apex/com.android.art/lib64/libart.so (art::gc::collector::GarbageCollector::Run(art::gc::GcCause, bool)+336)
  #04  pc 0x000000000048a434  /apex/com.android.art/lib64/libart.so (art::gc::Heap::CollectGarbageInternal(art::gc::collector::GcType, art::gc::GcCause, bool, unsigned int)+508)
  #05  pc 0x0000000000da4790  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+112)
  #06  pc 0x0000000000389d00  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (java.lang.Runtime.gc+160)
  #07  pc 0x0000000000328194  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #08  pc 0x00000000003262f8  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithVarArgs<_jmethodID*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, std::__va_list)+924)
  #09  pc 0x0000000000632850  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list)+180)
  #10  pc 0x0000000000044a04  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmono-android.release.so (_JNIEnv::CallVoidMethod(_jobject*, _jmethodID*, ...)+120) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #11  pc 0x0000000000041ab4  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmono-android.release.so (xamarin::android::internal::OSBridge::gc_cross_references(int, MonoGCBridgeSCC**, int, MonoGCBridgeXRef*)+228) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #12  pc 0x000000000029499c  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #13  pc 0x00000000002b0ee4  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #14  pc 0x00000000002adf80  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #15  pc 0x00000000002aef1c  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #16  pc 0x000000000029e108  /data/app/~~luR0FflQDpPlCz8M8OSAYQ==/MyApp-wEsw8Cu_Nu2KQa8R5dmkmw==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (mono_gc_collect+44) (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
```

```
  #00  pc 0x00000000000e088c  /apex/com.android.runtime/lib64/bionic/libc.so (__ioctl+12)
  #01  pc 0x000000000009765c  /apex/com.android.runtime/lib64/bionic/libc.so (ioctl+160)
  #02  pc 0x000000000056c41c  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::CopyIoctl(void*, void*, unsigned long, bool)+108)
  #03  pc 0x000000000056b734  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::MapMovingSpacePages(unsigned long, unsigned long, bool, bool)+496)
  #04  pc 0x00000000002d2910  /apex/com.android.art/lib64/libart.so (void art::gc::collector::MarkCompact::CompactMovingSpace<-1>(unsigned char*)+8384)
  #05  pc 0x000000000056972c  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::CompactionPhase()+176)
  #06  pc 0x0000000000565c10  /apex/com.android.art/lib64/libart.so (art::gc::collector::MarkCompact::RunPhases()+600)
  #07  pc 0x0000000000591b40  /apex/com.android.art/lib64/libart.so (art::gc::collector::GarbageCollector::Run(art::gc::GcCause, bool)+324)
  #08  pc 0x0000000000438220  /apex/com.android.art/lib64/libart.so (art::gc::Heap::CollectGarbageInternal(art::gc::collector::GcType, art::gc::GcCause, bool, unsigned int)+564)
  #09  pc 0x00000000003243c0  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+112)
  #10  pc 0x0000000000357860  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (java.lang.Runtime.gc+160)
  #11  pc 0x0000000000368774  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #12  pc 0x0000000000367168  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithVarArgs<art::ArtMethod*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, art::ArtMethod*, std::__va_list)+812)
  #13  pc 0x000000000072c0cc  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list)+192)
  #14  pc 0x0000000000044a04  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmono-android.release.so (_JNIEnv::CallVoidMethod(_jobject*, _jmethodID*, ...)+120) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #15  pc 0x0000000000041ab4  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmono-android.release.so (xamarin::android::internal::OSBridge::gc_cross_references(int, MonoGCBridgeSCC**, int, MonoGCBridgeXRef*)+228) (BuildId: 0ac0c2dff5c5d78ce61b1584a5559028fcb1eec3)
  #16  pc 0x000000000029499c  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #17  pc 0x00000000002b0ee4  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #18  pc 0x00000000002adf80  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #19  pc 0x00000000002aef1c  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #20  pc 0x000000000029e108  /data/app/~~sMlozcWZGzucyjy9nI9ffA==/MyApp-tmh1Tk4ZWhTGhyMHIqTVzQ==/split_config.arm64_v8a.apk!libmonosgen-2.0.so (mono_gc_collect+44) (BuildId: 1be7c7c5272957bf2e9a5560cbbed0b969688e5a)
  #
```

**Comment by b12kab** [comment]
I, like @fabEps, have way over 1/2 of my crashes sourced from Mono. I've also two [separate](https://github.com/dotnet/runtime/issues/116542) [issues](https://github.com/dotnet/runtime/issues/116541) that are specific to mono as well that have yet to be triaged.

I realize that these are very difficult to pin down; however we are getting penalized by our users and Google on these crashes. Some movement in these would be greatly appreciated.

**Comment by b12kab** [comment]
@steveisok ^^^

**Comment by steveisok** [comment]
> [@steveisok](https://github.com/steveisok) ^^^

Thanks for the nudge... I replied to both issues. You're likely going to need to bump to .NET 9 and see if the latest changes fix the problem.

**Comment by mnxamdev** [comment]
We're still seeing this in .Net 9, we've been on it for quite some time already. Our Android app is being put well above the bad behavior threshold because of this thus hindering our efforts to compete in the Google Play Store. Additionally, we're spending money with a firm to help us with our store presence and search but we're probably losing money on that because of this error that hasn't been fixed for close to a year.

**Comment by steveisok** [comment]
> We're still seeing this in .Net 9, we've been on it for quite some time already. Our Android app is being put well above the bad behavior threshold because of this thus hindering our efforts to compete in the Google Play Store. Additionally, we're spending money with a firm to help us with our store presence and search but we're probably losing money on that because of this error that hasn't been fixed for close to a year.

There is a chance the issue you are experiencing is not necessarily the same as this one. Can you provide a repro? If you can't describing the conditions leading up to the crash is your next best bet.

**Comment by mnxamdev** [comment]
> > We're still seeing this in .Net 9, we've been on it for quite some time already. Our Android app is being put well above the bad behavior threshold because of this thus hindering our efforts to compete in the Google Play Store. Additionally, we're spending money with a firm to help us with our store presence and search but we're probably losing money on that because of this error that hasn't been fixed for close to a year.
> 
> There is a chance the issue you are experiencing is not necessarily the same as this one. Can you provide a repro? If you can't describing the conditions leading up to the crash is your next best bet.

@steveisok we don't know what is causing this. If these traces aren't related to this issue then some insight from this limited information would be greatly appreciated:

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 26698 >>> com.myapp <<<

backtrace:
  #00  pc 0x000000000000ac48  /system/lib64/android.hidl.memory.token@1.0.so (android::hidl::memory::token::V1_0::BpHwMemoryToken::linkToDeath(android::sp<android::hardware::hidl_death_recipient> const&, unsigned long)+232)
  #01  pc 0x000000000022c620  /data/app/~~gWKzQh898lyxzu6__2ekfg==/com.myapp-zF2jZNthhAOTIp0TPYXLyA==/split_config.arm64_v8a.apk
```

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 9437 >>> com.myapp <<<

backtrace:
  #00  pc 0x00000000006664bc  /apex/com.android.art/lib64/libart.so (art::JNI<false>::GetObjectClass(_JNIEnv*, _jobject*)+176)
  #01  pc 0x0000000000020f10  /data/app/~~IVWsjOfUGh2SRREJ9Pqpmw==/com.myapp-vaoBejWv4xT1G0mmW_teow==/split_config.arm64_v8a.apk
```

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 1599 >>> com.myapp <<<

backtrace:
  #00  pc 0x0000000000355db8  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithJValues<art::ArtMethod*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, art::ArtMethod*, jvalue const*)+664)
  #01  pc 0x00000000007546b8  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodA(_JNIEnv*, _jobject*, _jmethodID*, jvalue const*)+108)
  #02  pc 0x0000000000021db0  /data/app/~~gF4pN_oAIqMj9k7XFhAyRw==/com.myapp-xj3L8jlsOGKIsWtRV6rr7g==/split_config.arm64_v8a.apk
```

**Comment by BrzVlad** [comment]
This issue is just a mix of unrelated crashes, some of which have been addressed. I think it is best to close this since it just adds confusion and it is currently not helpful .

Some relevant pointers
- try out the latest .NET 9 version, 9.0.7, since it has additional fixes
- if you are still getting crashes try out `<AndroidEnableSGenConcurrent>False</AndroidEnableSGenConcurrent>`, in order to disable concurrent gc
- if you are still getting crashes try out using new gc bridge. For this you would need to pass the following environment variable `MONO_GC_PARAMS=bridge-implementation=new`. You can do this that creating a new file containing this env variable definition (in `Platforms/Android/Environment.txt`) and then adding it to the csproj via `<AndroidEnvironment Include="Platforms\Android\Environment.txt" />` inside `ItemGroup`.
- if you are still encountering crashes, please submit new issue, add `os-android` label at least
   - crashes in libmonosgen-2.0.so are .NET runtime related. Any crash in the runtime will have this format, it doesn't mean the crashes are related with each other. Please symbolicate the crashes so we can see what could be the culprit. In order to do this, follow the instructions provided by Marek in this comment https://github.com/dotnet/runtime/issues/109921#issuecomment-2591887906
   - crashes that are not in the .NET runtime, like some of the above in `libart.so`, inside the android runtime GC, are not really actionable by us. You might want to report the issue to android runtime devs for example in this case.
